crypto/cmp/cmp_vfy.c

```diff
@@ -0,0 +1,868 @@
+ end:
+        sk_X509_pop_free(found_crts, X509_free);
+    }
+
+    return scrt;
+}
+
+/*
+ * Validate the protection of the given PKIMessage using either password-
+ * based mac (PBM) or a signature algorithm. In the case of signature algorithm,
+ * the certificate can be provided in ctx->srvCert,
+ * else it is taken from ctx->untrusted_certs (which should include extraCerts
+ * first) and from ctx->trusted and validated against ctx->trusted.
+ *
+ * If ctx->permitTAInExtraCertsForIR is true, the trust anchor may be taken from
+ * the extraCerts field when a self-signed certificate is found there which can
+ * be used to validate the enrolled certificate returned in IP.
+ *  This is according to the need given in 3GPP TS 33.310.
+ *
+ * returns 1 on success, 0 on error or validation failed
+ */
+int OSSL_CMP_validate_msg(OSSL_CMP_CTX *ctx, const OSSL_CMP_MSG *msg)
+{
+    X509_ALGOR *alg;
+    int nid = NID_undef, pk_nid = NID_undef;
+    const ASN1_OBJECT *algorOID = NULL;
+    X509 *scrt = NULL;
+
+    if (!ossl_assert(ctx != NULL && msg != NULL && msg->header != NULL))
+        return 0;
+
+    if ((alg = msg->header->protectionAlg) == NULL /* unprotected message */
+            || msg->protection == NULL || msg->protection->data == NULL) {
+        CMPerr(0, CMP_R_MISSING_PROTECTION);
+        return 0;
+    }
+
+    /* determine the nid for the used protection algorithm */
+    X509_ALGOR_get0(&algorOID, NULL, NULL, alg);
+    nid = OBJ_obj2nid(algorOID);
+
+    switch (nid) {
+        /* 5.1.3.1.  Shared Secret Information */
+    case NID_id_PasswordBasedMAC:
+        if (CMP_verify_PBMAC(msg, ctx->secretValue)) {
+            /*
+             * RFC 4210, 5.3.2: 'Note that if the PKI Message Protection is
+             * "shared secret information", then any certificate transported in
+             * the caPubs field may be directly trusted as a root CA
+             * certificate by the initiator.'
+             */
+            switch (ossl_cmp_msg_get_bodytype(msg)) {
+            case OSSL_CMP_PKIBODY_IP:
+            case OSSL_CMP_PKIBODY_CP:
+            case OSSL_CMP_PKIBODY_KUP:
+            case OSSL_CMP_PKIBODY_CCP:
+                if (ctx->trusted != NULL) {
+                    STACK_OF(X509) *certs = msg->body->value.ip->caPubs;
+                    /* value.ip is same for cp, kup, and ccp */
+
+                    if (!ossl_cmp_X509_STORE_add1_certs(ctx->trusted, certs, 0))
+                        /* adds both self-signed and not self-signed certs */
+                        break;
+                }
+            }
+            return 1;
+        }
+        break;
+
+        /* TODO: 5.1.3.2.  DH Key Pairs --> GitHub issue#87 */
+    case NID_id_DHBasedMac:
+        CMPerr(0, CMP_R_UNSUPPORTED_PROTECTION_ALG_DHBASEDMAC);
+        break;
+
+        /*
+         * 5.1.3.3.  Signature
+         * TODO: should that better white-list DSA/RSA etc.?
+         * -> check all possible options from OpenSSL, should there be macro?
```

Comment:
Outstanding TODO here?

Reply:
I removed the confusing parts of the comment.